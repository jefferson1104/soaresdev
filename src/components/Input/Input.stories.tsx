import { Story, Meta } from '@storybook/react/types-6-0'
import Input, { InputProps } from '.'

export default {
  title: 'Input',
  component: Input,
  args: {
    label: 'E-mail',
    name: 'email',
    initialValue: '',
    placeholder: 'john.doe@example.com'
  },
  argTypes: {
    onInput: { action: 'changed' },
    icon: { type: '' }
  }
} as Meta

export const Default: Story<InputProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <Input {...args} />
  </div>
)

export const withError: Story<InputProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <Input {...args} />
  </div>
)

withError.args = {
  error: 'Ops...something is wrong'
}
