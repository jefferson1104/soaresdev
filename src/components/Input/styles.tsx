import styled, { css, DefaultTheme } from 'styled-components'

import { InputProps } from '.'

type WrapperProps = Pick<InputProps, 'disabled'> & { error?: boolean }

const wrapperModifiers = {
  error: (theme: DefaultTheme) => css`
    ${InputWrapper} {
      border-color: ${theme.colors.red};
    }
    ${Label} {
      color: ${theme.colors.red};
    }
  `,
  disabled: () => css`
    ${Label},
    ${Input} {
      cursor: not-allowed;
    },

  `
}

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, error, disabled }) => css`
    ${error && wrapperModifiers.error(theme)}
    ${disabled && wrapperModifiers.disabled()}
  `}
`

export const InputWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    background: ${theme.colors.darkGray};
    border-radius: ${theme.border.radius};
    padding: 0 ${theme.spacings.xsmall};
    border: 0.2rem solid;
    border-color: ${theme.colors.secondary};
    transition: ${theme.transition.default};

    &:focus-within {
      border: 2px solid ${theme.colors.primary};
    }
  `}
`

export const Input = styled.input`
  ${({ theme }) => css`
    color: ${theme.colors.white};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} 0;

    background: transparent;
    border: 0;
    outline: none;
    width: 100%;

    &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0 ${theme.spacings.small}
        ${theme.colors.lightGray} inset;
      filter: none;
    }
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.medium};
    color: ${theme.colors.white};
    cursor: pointer;
  `}
`

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.red};
    font-size: ${theme.font.sizes.xsmall};
  `}
`
