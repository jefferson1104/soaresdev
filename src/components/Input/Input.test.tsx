import { render, screen, waitFor } from 'utils/test-utils'
import userEvent from '@testing-library/user-event'

import Input from '.'

describe('Input component', () => {
  it('Renders with Label', () => {
    render(<Input label="Label" name="label" />)

    const label = screen.getByLabelText('Label')

    expect(label).toBeInTheDocument()
  })

  it('Renders without Label', () => {
    render(<Input />)

    const label = screen.queryByLabelText('Label')

    expect(label).not.toBeInTheDocument()
  })

  it('Renders with placeholder', () => {
    render(<Input placeholder="hey you" />)

    const placeholder = screen.getByPlaceholderText('hey you')

    expect(placeholder).toBeInTheDocument()
  })

  it('Renders with error', () => {
    const { container } = render(
      <Input label="testField" error="Error message" />
    )

    expect(screen.getByText('Error message')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('Is accessible by tab', () => {
    render(<Input label="TextField" name="TextField" />)

    const input = screen.getByLabelText('TextField')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).toHaveFocus()
  })

  it('Is not accessible by tab when disabled', () => {
    render(<Input label="TextField" name="TextField" disabled />)

    const input = screen.getByLabelText('TextField')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).not.toHaveFocus()
  })

  it('Changes its value when typing', async () => {
    const onInputChange = jest.fn()

    render(
      <Input onInputChange={onInputChange} label="TextField" name="TextField" />
    )

    const input = screen.getByRole('textbox')
    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInputChange).toHaveBeenCalledTimes(text.length)
    })

    expect(onInputChange).toHaveBeenCalledWith(text)
  })

  it('Does not changes its value when disabled', async () => {
    const onInputChange = jest.fn()
    render(
      <Input
        onInputChange={onInputChange}
        label="TextField"
        name="TextField"
        disabled
      />
    )

    const input = screen.getByRole('textbox')
    expect(input).toBeDisabled()

    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).not.toHaveValue(text)
    })
    expect(onInputChange).not.toHaveBeenCalled()
  })
})
