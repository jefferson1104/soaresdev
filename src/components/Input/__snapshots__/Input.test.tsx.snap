// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component Renders with error 1`] = `
.c0 .c3 {
  border-color: #e74c3c;
}

.c0 .c1 {
  color: #e74c3c;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #272530;
  border-radius: 0.4rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #2E2F42;
  -webkit-transition: 0.5s all ease-in-out;
  transition: 0.5s all ease-in-out;
}

.c4:focus-within {
  border: 2px solid #4bbecf;
}

.c5 {
  color: #FAFAFA;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.8rem;
  padding: 0.8rem 0;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c5:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4rem #b3b3b3 inset;
  -webkit-filter: none;
  filter: none;
}

.c2 {
  font-size: 1.8rem;
  color: #FAFAFA;
  cursor: pointer;
}

.c6 {
  color: #e74c3c;
  font-size: 1.4rem;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
  >
    testField
  </label>
  <div
    class="c3 c4"
  >
    <input
      class="c5"
      type="text"
      value=""
    />
  </div>
  <p
    class="c6"
  >
    Error message
  </p>
</div>
`
